src/CollectionInterface.php

```diff
@@ -70,4 +77,105 @@ public function getType(): string;
+     */
+    public function column(string $propertyOrMethod): array;
+
+    /**
+     * Return the first item of the collection
+     *
+     * @return mixed
+     */
+    public function first();
+
+    /**
+     * Returns the last item of the collection
+     *
+     * @return mixed
+     */
+    public function last();
+
+    /**
+     * Sort the collection by property or method with given sort order
+     *
+     * This will always leave the original collection untouched and will return a new one.
+     *
+     * @param string $propertyOrMethod
+     * @param string $order
+     *
+     * @return CollectionInterface
+     * @throws InvalidSortOrderException Raised if neither ASC nor DESC was given
+     */
+    public function sort(string $propertyOrMethod, string $order = self::SORT_ASC): self;
+
+    /**
+     * Filter out items of the collection which don't math the criteria of given callback.
+     *
+     * This will always leave the original collection untouched and will return a new one.
+     *
+     * @param callable $callback
+     *
+     * @return CollectionInterface
+     */
+    public function filter(callable $callback): self;
+
+    /**
+     * Create a new collection where items match the criteria of given callback
+     *
+     * This will always leave the original collection untouched and will return a new one.
+     *
+     * @param string $propertyOrMethod
+     * @param mixed  $value
+     *
+     * @return CollectionInterface
+     */
+    public function where(string $propertyOrMethod, $value): self;
+
+    /**
+     * Apply a given callback method on each item of the collection
+     *
+     * This will always leave the original collection untouched and will return a new one.
+     *
+     * @param callable $callback
+     *
+     * @return CollectionInterface
+     */
+    public function map(callable $callback): self;
+
+    /**
+     * Create a new collection with divergent items between current and given collection
+     *
+     * @param CollectionInterface $other
+     *
+     * @return CollectionInterface
+     * @throws DiverseCollectionException Raised if given collection is not of same type
+     */
+    public function diff(CollectionInterface $other): self;
+
+    /**
+     * Create a new collection with intersecting item between current and given collection
+     *
+     * @param CollectionInterface $other
+     *
+     * @return CollectionInterface
+     * @throws DiverseCollectionException Raised if given collection is not of same type
+     */
+    public function intersect(CollectionInterface $other): self;
+
+    /**
+     * Merge current items and items of given collections into a new one
+     *
+     * @param CollectionInterface[] ...$collections
+     *
+     * @return CollectionInterface
+     * @throws DiverseCollectionException Raised if one of the collection is not of same type
+     */
+    public function merge(...$collections): self;
```

Comment:
Why doesn't this have a `CollectionInterface` type?

Reply:
This ofc is a good question. Will change the docblocks and typehints there.